RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitSoAProducer.cc

```diff
@@ -32,27 +35,34 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
       }
     }
 
-    void produce(edm::StreamID, device::Event& event, const device::EventSetup& setup) const override {
+    void acquire(device::Event const& event, const device::EventSetup& setup) override {
       const typename CAL::TopologyTypeDevice& topology = setup.getData(topologyToken_);
 
       uint32_t num_recHits = 0;
       for (const auto& token : recHitsToken_)
         num_recHits += event.get(token.first)->metadata().size();
 
-      reco::PFRecHitDeviceCollection pfRecHits{(int)num_recHits, event.queue()};
+      pfRecHits_.emplace((int)num_recHits, event.queue());
+      *size_ = 0;
 
       if (num_recHits != 0) {
         PFRecHitProducerKernel<CAL> kernel{event.queue(), num_recHits};
         for (const auto& token : recHitsToken_)
           kernel.processRecHits(
-              event.queue(), event.get(token.first), setup.getData(token.second), topology, pfRecHits);
-        kernel.associateTopologyInfo(event.queue(), topology, pfRecHits);
+              event.queue(), event.get(token.first), setup.getData(token.second), topology, *pfRecHits_);
+        kernel.associateTopologyInfo(event.queue(), topology, *pfRecHits_);
+        auto size_d = cms::alpakatools::make_device_view<uint32_t>(event.queue(), pfRecHits_->view().size());
+        alpaka::memcpy(event.queue(), size_, size_d);
       }
+    }
+
+    void produce(device::Event& event, const device::EventSetup& setup) override {
+      event.emplace(pfRecHitsToken_, std::move(*pfRecHits_));
+      event.emplace(sizeToken_, *size_);
+      pfRecHits_.reset();
 
       if (synchronise_)
         alpaka::wait(event.queue());
```

Comment:
The comment of the `synchronise` configuration parameter says
> Add synchronisation point after execution (for benchmarking asynchronous execution)

so maybe this code should be moved to the end of `acquire()`? (as no asynchronous activities are launched from the `produce()`)

Reply:
Mhm, ok.